jdaviz/configs/imviz/helper.py

```diff
@@ -195,6 +199,193 @@ def link_data(self, **kwargs):
+            if a Numpy array is given.**
+
+        Returns
+        -------
+        bad_regions : list of (obj, str)
+            List of ``(region, reason)`` tuple for region objects that failed to load.
+            If all the regions loaded successfully, this will be empty.
+
+        """
+        from photutils.aperture import (CircularAperture, SkyCircularAperture,
+                                        EllipticalAperture, SkyEllipticalAperture,
+                                        RectangularAperture, SkyRectangularAperture)
+        from regions import (CirclePixelRegion, CircleSkyRegion,
+                             EllipsePixelRegion, EllipseSkyRegion,
+                             RectanglePixelRegion, RectangleSkyRegion)
+        from jdaviz.core.region_translators import regions2roi, aperture2regions
+
+        # If user passes in one region obj instead of list, try to be smart.
+        if not isinstance(regions, (list, tuple, Regions)):
+            regions = [regions]
+
+        n_loaded = 0
+        bad_regions = []
+
+        # To keep track of masked subsets.
+        msg_prefix = 'MaskedSubset'
+        msg_count = _next_subset_num(msg_prefix, self.app.data_collection.subset_groups)
+
+        # Subset is global but reference data is viewer-dependent.
+        if refdata_label is None:
+            data = self.default_viewer.state.reference_data
+        else:
+            data = self.app.data_collection[refdata_label]
+        has_wcs = data_has_valid_wcs(data)
+
+        for region in regions:
+            if isinstance(region, (SkyCircularAperture, SkyEllipticalAperture,
+                                   SkyRectangularAperture, CircleSkyRegion,
+                                   EllipseSkyRegion, RectangleSkyRegion)) and not has_wcs:
+                bad_regions.append((region, 'Sky region provided but data has no WCS'))
+                continue
+
+            # photutils: Convert to regions shape first
+            if isinstance(region, (CircularAperture, SkyCircularAperture,
+                                   EllipticalAperture, SkyEllipticalAperture,
+                                   RectangularAperture, SkyRectangularAperture)):
+                region = aperture2regions(region)
+
+            # regions: Convert to ROI.
+            # NOTE: Out-of-bounds ROI will succeed; this is native glue behavior.
+            if isinstance(region, (CirclePixelRegion, CircleSkyRegion,
+                                   EllipsePixelRegion, EllipseSkyRegion,
+                                   RectanglePixelRegion, RectangleSkyRegion)):
+                state = regions2roi(region, wcs=data.coords)
+
+                # TODO: Do we want user to specify viewer? Does it matter?
+                self.app.session.edit_subset_mode._mode = NewMode
+                self.default_viewer.apply_roi(state)
+                self.default_viewer.session.edit_subset_mode.edit_subset = None  # No overwrite next iteration # noqa
```

Comment:
does this need to be reverted after the API call to whatever the user had set? Or is there a way to load these regions directly into the state without having to fake them as drawn regions?

Reply:
> does this need to be reverted after the API call to whatever the user had set? 

I don't think so? If user is using the API to load a bunch of Subset regions, and then user wants to interactively draw one, chances are user wants to draw a new one.

> is there a way to load these regions directly into the state without having to fake them as drawn regions?

`apply_roi` is the only way I know how to do this. Ideas welcome!

Reply:
My only idea would be if we could just add directly to `app.data_collection.subset_groups`, but I'm not sure if that's writeable or not.

Reply:
Based on the implementation here, I cannot find a setter for the public property. In theory, you can write to almost anything in Python but since that would involve touching the private attribute, maybe we shouldn't go that route?

https://github.com/glue-viz/glue/blob/578369fa09046471c32fd9c0132961474f678248/glue/core/data_collection.py#L390-L395